### PR TITLE
feat: gaps for external executor

### PR DESCRIPTION
**Pull request by AdiGajbhiye**
## Overview

- check connectivity with backend before running operation
- add telemetry events for external executor

## Checklist

- [ ] I have run this code and it appears to resolve the stated issue
- [ ] `README.md` updated and added information about my change

### CODE REVIEW

File: src/dbt_client/dbtCloudIntegration.ts [hunk 1]
```diff
@@ -137,13 +138,36 @@ export class DBTCloudProjectIntegration
       this.rebuildManifestDiagnostics,
       this.pythonBridgeDiagnostics,
     );
+    this.checkConnectivity();
+  }
+
+  throwIfNotAuthenticated() {
```
Comment (mdesmet) [inline src/dbt_client/dbtCloudIntegration.ts:144]: make method private

File: src/dbt_client/dbtCloudIntegration.ts [hunk 2]
```diff
@@ -137,13 +138,36 @@ export class DBTCloudProjectIntegration
...
+    if (!this.isAuthenticated) {
+      throw new Error("Unable to connect with altimate backend");
```
Comment (mdesmet) [inline src/dbt_client/dbtCloudIntegration.ts:146]: error message might be misleading.

File: src/dbt_client/dbtCloudIntegration.ts [hunk 3]
```diff
@@ -395,6 +412,8 @@ export class DBTCloudProjectIntegration
   }
 
   async unsafeCompileQuery(query: string): Promise<string | undefined> {
+    this.telemetry.sendTelemetryEvent("dbtCloud:unsafeCompileQuery");
+    this.throwIfNotAuthenticated();
```
Comment (mdesmet) [inline src/dbt_client/dbtCloudIntegration.ts:416]: I think this should work differently. we should send an event in dbtProject and add a projectIntegration dimension. that way we can easily compare events for both cloud vs core.

File: src/validation_provider/index.ts [hunk 4]
```diff
@@ -20,28 +21,31 @@ export class ValidationProvider implements Disposable {
...
     );
   }
 
-  private async validateCredentials() {
-    const config = this.altimate.getConfig();
-    if (!config) {
-      return;
-    }
-    const { key, instance } = config;
+  async validateCredentials(silent: boolean) {
```
Comment (mdesmet) [inline src/validation_provider/index.ts:32]: Create method alias for silent mode, booleans are difficult to understand in the calling code.

File: src/validation_provider/index.ts [hunk 5]
```diff
@@ -20,28 +21,31 @@ export class ValidationProvider implements Disposable {
           if (!e.affectsConfiguration("dbt")) {
             return;
           }
-          this.validateCredentials();
+          this.validateCredentials(false);
```
Comment (mdesmet) [inline src/validation_provider/index.ts:24]: it will work differently from before as it was showing an error message. Is that intentional?

Comment (AdiGajbhiye) [reply src/validation_provider/index.ts]: not sure what you mean. i am still getting error notification if credentials are wrong.

File: src/validation_provider/index.ts [hunk 6]
```diff
@@ -28,20 +29,31 @@ export class ValidationProvider implements Disposable {
...
-    const { key, instance } = config;
+  validateCredentials() {
+    this._validateCredentials(false);
+  }
+
+  validateCredentialsSilently() {
+    this._validateCredentials(true);
+  }
+
+  private async _validateCredentials(silent: boolean) {
```
Comment (mdesmet) [inline src/validation_provider/index.ts:40]:
```suggestion
  private async validateCredentials(silent: boolean = false) {
```

File: src/validation_provider/index.ts [hunk 7]
```diff
@@ -9,6 +9,7 @@ export class ValidationProvider implements Disposable {
   private disposables: Disposable[] = [];
   private currInstanceName: string | undefined;
   private currAPIKey: string | undefined;
+  isAuthenticated = false;
```
Comment (mdesmet) [inline src/validation_provider/index.ts:12]: maybe it is better to provide a method for this and make this private?

**Review by mdesmet — APPROVED** [review]

**Pull request merged**